File: decorator.go
package main

import (
	"decorator"
	"fmt"
)

func main() {
	gift := decorator.NewGift("flower")
	gift_packed_first := decorator.NewPack(gift, "red ribbon")
	gift_packed_second := decorator.NewPack(gift_packed_first, "box")

	fmt.Println("a good sold, its detail:" + gift_packed_second.GetDesc()) // a good sold, its detail:flower, with red ribbon packed, with box packed
}

File: java/src/pattern/Decorator.java
package pattern;

import pattern.decorator.*;

public class Decorator {
    public static void main(String[] args){
        Toy toyCar = new Toy("toy car");
        System.out.println(toyCar.sold());

        Pack box = new Pack("box");
        box.packet(toyCar);
        System.out.println(box.sold());

        Pack redRibbon = new Pack("red ribbon");
        redRibbon.packet(box);
        System.out.println(redRibbon.sold());
    }
}
